macaroon.js

```diff
@@ -2,30 +2,267 @@
 
 const sjcl = require('sjcl');
 const nacl = require('tweetnacl');
+const textEncoding = require('text-encoding');
 nacl.util = require('tweetnacl-util');
-const varint = require('varint');
+const utf8Encoder = new textEncoding.TextEncoder('utf-8');
+const utf8Decoder = new textEncoding.TextDecoder('utf-8', {fatal: true});
 
 const NONCELEN = 24;
 
-const V2_TYPES = {
-  LOCATION: 1,
-  IDENTIFIER: 2,
-  VID: 4,
-  SIGNATURE: 6
+const FIELD_EOS = 0;
+const FIELD_LOCATION = 1;
+const FIELD_IDENTIFIER = 2;
+const FIELD_VID = 4;
+const FIELD_SIGNATURE = 6;
+
+/**
+ * Convert a string to a Uint8Array by utf-8
+ * encoding it.
+ * @param {string} s The string to convert.
+ * @return {Uint8Array} The resulting bytes.
+ */
+const stringToBytes = function(s) {
+  if (s === null) {
+    return null;
+  }
+  return utf8Encoder.encode(s);
 };
 
-// Uses protobuf-style varints
-function toVarintBuf(number) {
-  return Buffer.from(varint.encode(number));
-}
+/**
+ * Convert a Uint8Array to a string by
+ * utf-8 decoding it. Throws an exception if
+ * the bytes do not represent well-formed utf-8.
+ * @param {Uint8Array} The bytes to convert.
+ * @return {string} The resulting string.
+ */
+const bytesToString = function(b) {
+  if (b === null) {
+    return null;
+  }
+  return utf8Decoder.decode(b);
+};
+
+/**
+ * The maximum integer that can be manipulated with
+ * JS bitwise operations.
+ */
+const maxInt = Math.pow(2, 32)-1;
+
+const ByteBuffer = class ByteBuffer {
+    /**
+     * Create a new ByteBuffer. A ByteBuffer holds
+     * a Uint8Array that it grows when written to.
+     * @param {int} The initial capacity of the buffer.
```

Comment:
JSDoc usually is in the format `@param {type} name Description`.
So in this case `@param {int} capacity The initial capacity of the buffer.`